Network/Socket/Shutdown.hs

```diff
@@ -34,3 +40,40 @@ shutdown s stype = void $ withFdSocket s $ \fd ->
 
 foreign import CALLCONV unsafe "shutdown"
   c_shutdown :: CInt -> CInt -> IO CInt
+
+
+-- | Closing a socket gracefully.
+--   This sends TCP FIN and check if TCP FIN is received from the peer.
+--   The second argument is time out to receive TCP FIN in millisecond.
+--   In both normal cases and error cases, socket is deallocated finally.
+--
+--   Since: 3.1.1.0
+gracefulClose :: Socket -> Int -> IO ()
+gracefulClose s tm = (sendRecvFIN `E.finally` close s) `E.catch` ignore
+  where
+    sendRecvFIN = do
+        -- Sending TCP FIN.
+        shutdown s ShutdownSend
+        -- Waiting TCP FIN.
+        E.bracket (mallocBytes bufSize) free recvEOF
+    recvEOF buf = loop 0
+      where
+        loop delay = do
+            -- We don't check the (positive) length.
+            -- In normal case, it's 0. That is, only FIN is received.
+            -- In error cases, data is available. But there is no
+            -- application which can read it. So, let's stop receiving
+            -- to prevent attacks.
+            r <- recvBufNoWait s buf bufSize
+            let delay' = delay + clock
```

Comment:
Here you're counting the time you _asked_ to sleep, not the time that was actually slept (which can be a lot more when the system is busy).

I think _if_ we implement loop-based sleeping (which we may not have to do based on @snoyberg's comment), we should re-check the time that has passed.

Reply:
@nh2 We have to use `threadDelay` on WIndows due to IO manager's limitation. Though your point is absolutely right, `network` should not depends on `time` nor `hourglass` unfortunately. Do you have any ideas to check the time using `base` only?